dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorLKMM.java

```diff
@@ -264,15 +253,60 @@ public List<Event> visitRMWXchg(RMWXchg e) {
     public List<Event> visitLKMMLock(LKMMLock e) {
         Register dummy = e.getThread().newRegister(types.getArchType());
         Expression zero = expressions.makeZero(dummy.getType());
+
+        Load lockRead = newLockRead(e.getLock(), dummy);
         // In litmus tests, spin locks are guaranteed to succeed, i.e. its read part gets value 0
-        Load lockRead = Linux.newLockRead(dummy, e.getLock());
-        Event middle = e.getThread().getProgram().getFormat().equals(LITMUS) ?
+        Event checkLockValue = e.getThread().getProgram().getFormat().equals(LITMUS) ?
                 newAssume(expressions.makeEQ(dummy, zero)) :
                 newJump(expressions.makeNEQ(dummy, zero), (Label) e.getThread().getExit());
         return eventSequence(
                 lockRead,
-                middle,
-                Linux.newLockWrite(lockRead, e.getLock())
+                checkLockValue,
+                newLockWrite(lockRead, e.getLock())
         );
     }
+
+    // ============================== Helper methods to lower LKMM events to core events ===========================
+    /*
+        The following helper methods are used to generate core-level events with additional metadata attached,
+        for example, with custom printing capabilities.
+     */
+
+    private static Fence newCoreMemoryBarrier() {
+        return newFence(Tag.Linux.MO_MB);
+    }
+
+    private static Load newCoreLoad(Register reg, Expression addr, String mo) {
+        Load load = EventFactory.newLoadWithMo(reg, addr, mo);
+        load.setMetadata(LKMMLoad.CUSTOM_CORE_PRINTING);
+        return load;
+    }
+
+    private static Store newCoreStore(Expression addr, Expression value, String mo) {
+        Store store = EventFactory.newStoreWithMo(addr, value, mo);
+        store.setMetadata(LKMMStore.CUSTOM_CORE_PRINTING);
+        return store;
+    }
+
+    private static Load newLockRead(Expression lock, Register dummy) {
```

Comment:
For consistency with all other `Load`s, the register should be the first parameter